채널 부분 구현, axios 설정, 채널 api 이용하는 부분을 service 코드로 분리

hansanguk0222 opened the pull request:
# 오늘 한 일
## 채널 부분 구현
- 사이드바를 useEffect를 통하여 로그인 하는 순간 채널의 정보를 불러오도록 구현
- 클릭한 정보는 사이드 바를 줄여도 살아있게 구현(색깔은 아직 넣지 못함, 좋은 방법이 있을까요ㅜㅜ)
- 호버하면 색깔 변함
![Peek 2020-11-25 23-44](https://user-images.githubusercontent.com/50985451/100242473-29df1980-2f78-11eb-9108-ca383d446d2d.gif)


## axios 설정
- 아까 설정한 값 그대로 넣었습니다. 충돌이 나면 빨리 해결할 수 있을 듯 합니다.

## service 코드 작성
- saga에서 axios모듈을 직접 이용하였는데 service 코드로 분리하니 간결하고 좋았습니다.

hansanguk0222 committed:
[feat] Request 하기 전 axios에서 헤더의 토큰과 공통 URL을 사용할 수 있도록 세팅

hansanguk0222 committed:
[feat] 좌측 사이드바에 들어갈 채널 선택 부분을 구현(채널 추가 버튼, 검색 버튼은 미구현)

hansanguk0222 committed:
[feat] api를 이용하여 채널의 정보를 불러오는 작업을 service 코드로 분리

rnjshippo commented on client/src/components/ChannelListBox/ChannelList/Channel/Channel.tsx:
```diff
@@ -0,0 +1,53 @@
+    name: state.channels.channelList[args.idx].name,
+    isPublic: state.channels.channelList[args.idx].isPublic,
+    current: state.channels.current,
+  }));
```
이 부분을 커스텀 훅으로 대체해봅시다! 제 pr에서 훅쓴거 참고하시면 좋을듯해요 ( hooks/useAuth.ts )

hansanguk0222 replied:
오 알겟습니다

hansanguk0222 replied:
완료했습니다ㅎㅎ

rnjshippo commented on client/src/components/ChannelListBox/ChannelList/Channel/Channel.tsx:
```diff
@@ -0,0 +1,53 @@
+  const onClick = useCallback(() => {
+    dispatch(onChangeCurrent(id));
+  }, []);
+
```
이 부분을 제대로 메모이제이션 하려면 두 번째 값으로 빈 배열이 아니라 [dispatch]가 들어가야 dispatch 함수가 바뀌어도 올바르게 dispatch를 할 수 있습니다!

그리고

useCallback, useMemo, ... 등등 여러 메모이제이션 기법들을 매번 쓸필요는 없다고 생각해요. 연산이 큰 경우에 적절하게 활용하면 성능 향상을 기대할 수 있지만, 여기저기 계속쓰면 메모리 낭비로 이어지고 역으로 성능에 좋지 않을 수 있어요!
아래 링크를 읽어보시면 도움이 되실 것 같습니다~

https://ideveloper2.dev/blog/2019-06-14--when-to-use-memo-and-use-callback/

hansanguk0222 replied:
해결했습니다ㅎㅎ

rnjshippo commented on client/src/components/ChannelListBox/ChannelList/ChannelList.tsx:
```diff
@@ -0,0 +1,54 @@
+
+  const callAPI = useCallback(() => {
+    dispatch(loadChannelsRequest());
+  }, []);
```
위에서 언급한것과 같은 맥락입니다! 메모이제이션을 아예 안하거나, 두번째 인자로 dispatch를 넣어야합니다

hansanguk0222 replied:
해결했습니다ㅎㅎ

rnjshippo commented on client/src/components/ChannelListBox/ChannelList/ChannelList.tsx:
```diff
@@ -0,0 +1,54 @@
+  }, []);
+
+  return (
+    <div>
```
최상위 요소가 단순히 div로 묶어주기만 한다면 div 보다는 <></> 로 감싸시는걸 추천~

hansanguk0222 replied:
해결했습니다ㅎㅎ

rnjshippo commented on client/src/components/ChannelListBox/ChannelListHeader/ChannelListHeader.tsx:
```diff
@@ -0,0 +1,66 @@
+  return (
+    <ChannelListHeaderWrapper>
+      <SubWrapper>
+        <ButtonWrapper onClick={onClick}>{showList ? '▽' : '▷'}</ButtonWrapper>
```
단순히 button이라면 이름을 그냥 Button해도 될듯해요! styeld 컴포넌트는 짧게 쓰는게 가독성이 좋아보이더라구요. 이것도 컨벤션으로 같이 얘기해보면 좋을 듯해요

hansanguk0222 replied:
이따가 상의 후 점심먹기 전 수정하겠습니다

rnjshippo commented on client/src/store/modules/channels.ts:
```diff
@@ -55,5 +53,7 @@ export const {
   loadChannelsFalse,
+  onChangeCurrent,
+  onChangeShowList,
 } = channelsSlice.actions;
```
이 부분 액션 내보낼때마다 추가하는게 귀찮고, 액션이 많아지면 관리 하기가 번잡할 것 같아요!

저는 아래 방식으로 내보내고 받는 쪽에서 AUTH_ACTIONS.액션이름 -> 이렇게 사용했는데 이것도 같이 컨벤션 정해봐요
```js
export const AUTH_ACTIONS = authSlice.actions;
```

hansanguk0222 replied:
이따가 상의 후 점심먹기 전 수정하겠습니다

rnjshippo commented on client/src/store/sagas/channelSaga.ts:
```diff
@@ -1,25 +1,21 @@
 import { all, fork, takeEvery, call, put } from 'redux-saga/effects';
-import axios from 'axios';
+import { channelsService } from '@/services/channels.service';
```
적용 감사합니다! 👍

hansanguk0222 replied:
이걸로 작업이 편해졌습니다ㅎㅎ

rnjshippo reviewed (commented):
인터렉티브한걸 벌써 끝내주시다니.. 고생많으셨습니다! 👏

rnjshippo commented on client/src/components/ChannelListBox/ChannelList/ChannelList.tsx:
```diff
@@ -0,0 +1,54 @@
+    showList: state.channels.showList,
+    current: state.channels.current,
+  }));
+
```
사가를 읽어보니 채널리스트 fetch에 실패했을 때 payload로 err를 전달하고 있으시네요.

그러면 state.show에는 에러 객체가 통째로 들어가는데, 지금 여기서 show 값으로 에러 핸들링하는 부분이 없는 것 같아요! 

fetching에 실패했을때의 처리를 해주시면 좋을 듯 합니다.

( 아마 401이라서 login페이지로 redirect만 시켜도 되지 않을 까 싶어요)

hansanguk0222 committed:
[feat] channel에서 state를 가져오는 부분을 훅으로 제작

hansanguk0222 committed:
[refactor]
1. 커스텀 훅으로 state를 가져옴
2. 쓸데없이 <div>로 감싸던 것을 <>를 이용하여 감쌈
3. useCallback이 불필요하게 사용되어 삭제

hansanguk0222 committed:
[refactor]
1. 채널 누르면 색깔 바뀌게 구현
2. 커스텀 훅으로 state를 관리

hansanguk0222 committed:
[refactor] 커스텀 훅으로 state를 관리

hansanguk0222 committed:
[refactor] loadChannelFalse에서 loadChannelFailure로 함수명을 변경

hansanguk0222 committed:
[refactor] useEffect에서 dispatch를 감지하여 dispatch에 변화가 았을 때 함수 컴포넌트에서 사이드 이팩트가 일어나도록 반영함

hansanguk0222 committed:
Revert "[refactor] useEffect에서 dispatch를 감지하여 dispatch에 변화가 았을 때 함수 컴포넌트에서 사이드 이팩트가 일어나도록 반영함"

This reverts commit aa1cb53527727388e594c3647e78837f66707722.

hansanguk0222 committed:
[refactor] useEffect에서 dispatch를 감지하게 하여 dispatch가 변하면 항상 최신 상태를 가르키도록 만듬

hansanguk0222 committed:
[refactor] 사용하지 않는 함수 제거

hansanguk0222 committed:
[refactor] 함수명 변경과 액션만 반환해주는 함수를 구현

hansanguk0222 committed:
[refactor]
1. 제일 작은 컴포넌트 이름을 channel -> channelItem로 변경
2. 사용하지 않는 함수 삭제

hansanguk0222 committed:
[refactor]
1. 파일명을 Channel -> ChannelItem으로 변경
2. 컴포넌트명을 Channel -> ChannelItem으로 변경

hansanguk0222 committed:
[refactor] 사용하지 않는 함수를 삭제

hansanguk0222 committed:
[refactor]
1. state의 이름을 변경
2. 사용하지 않는 함수를 삭제

hansanguk0222 committed:
[refactor] state의 이름을 channels -> channel로 변경
